rewiringamerica/embed.rewiringamerica.org
# Add modal glossary of payment methods

## Description

There's a lot going on here, so here's the high-level tour:

- The glossary modal itself looks (I think) exactly as designed. I've
  made a few edits to the copy compared to what's in the Figma; I've
  put comments in the Figma to explain why.

- Most of the logic of displaying a modal dialog is hand-written,
  though I added third-party packages to trap focus within it and stop
  the body from scrolling while it's up.

  I did try using Headless UI's `Dialog` component, which does all
  that out of the box, but (like the rest of Headless!) it's unusable
  with shadow DOM. It renders the dialog at the root of `document`,
  where it does not get any of the shadow DOM's styling; there seems
  to be no way to change this.

- The entire incentive-type chip is now a button. (The alternative was
  to make only the icon the button, which makes for an uncomfortably
  small click target.) This required a bit of refactoring of the
  incentive card component.

### Accessibility

- Keyboard navigation:

  - When you activate one of the chip buttons, focus automatically
    goes to the section header of the appropriate section. (The entire
    header is an expand/collapse button, not just the circled
    chevron.)

  - The modal traps focus while it's up.

  - Hitting Escape while the modal is up will dismiss it.

  - When the modal closes, focus returns to the chip button that
    caused it to open.

- Screenreader experience: I think it's OK? I am not an expert screen
  reader user by any means. It's possible a few more things could use
  ARIA roles.

  One judgment call I made is to aria-label the question icon in the
  chips as "show glossary". When a chip is focused, the reader says
  "Tax credit show glossary, button". It seemed good to have a clue as
  to what will happen if you click the button?

### Next steps

If this all seems suitable, I'll request translations.

## Test Plan

- Look at the component on all widths, in several browsers (tried
  Chrome, Firefox, Safari, MobileSafari). Make sure the section
  corresponding to the incentive type you clicked on is open as soon
  as the modal appears, and no others are open. Click on others to
  open and close. Open all the sections and make sure the scrolling is
  correct.

- Make sure the modal is dismissible by clicking away from the body,
  clicking the X button, or hitting Escape.

- Scroll-wheel with the cursor away from the modal body, and make sure
  the document doesn't scroll underneath.

- Navigate with keyboard: tab to the incentive-type chip and hit
  Enter. Make sure focus is immediately on the right section
  header. Tab and shift-tab and make sure focus is trapped within the
  modal. Hit Esc, or "enter" on the X button, and make sure focus is
  back on the original chip.

- Do all of the above with VoiceOver on.
